cpp/include/raft/neighbors/detail/ann_mg.cuh

```diff
@@ -0,0 +1,712 @@
+      }
+  }
+
+  // MG index deserialization
+  ann_mg_index(const raft::resources& handle,
+               const std::string& filename) {
+      std::ifstream is(filename, std::ios::in | std::ios::binary);
+      if (!is) { RAFT_FAIL("Cannot open file %s", filename.c_str()); }
+
+      mode_ = (raft::neighbors::mg::dist_mode)deserialize_scalar<int>(handle, is);
+      root_rank_ = 0;
+      num_ranks_ = deserialize_scalar<int>(handle, is);
+      dev_ids_.resize(num_ranks_);
+      std::iota(std::begin(dev_ids_), std::end(dev_ids_), 0);
+      nccl_comms_.resize(num_ranks_);
+
+      init_device_resources();
+      init_nccl_clique();
+
+      for (int rank = 0; rank < num_ranks_; rank++) {
+        RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+        auto& ann_if = ann_interfaces_.emplace_back();
+        ann_if.deserialize(dev_resources_[rank], is);
+      }
+
+      is.close();
+  }
+
+  ann_mg_index(const ann_mg_index&)                    = delete;
+  ann_mg_index(ann_mg_index&&)                         = default;
+  auto operator=(const ann_mg_index&) -> ann_mg_index& = delete;
+  auto operator=(ann_mg_index&&) -> ann_mg_index&      = default;
+
+  void init_device_resources() {
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+      dev_resources_.emplace_back();
+    }
+  }
+
+  void init_nccl_clique() {
+    RAFT_NCCL_TRY(ncclCommInitAll(nccl_comms_.data(), num_ranks_, dev_ids_.data()));
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+      raft::comms::build_comms_nccl_only(&dev_resources_[rank], nccl_comms_[rank], num_ranks_, rank);
+    }
+  }
+
+  void destroy_nccl_clique() {
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      cudaSetDevice(dev_ids_[rank]);
+      ncclCommDestroy(nccl_comms_[rank]);
+    }
+  }
+
+  void build(const ann::index_params* index_params,
+             raft::host_matrix_view<const T, IdxT, row_major> index_dataset)
+  {
+    if (mode_ == INDEX_DUPLICATION) {
+      for (int rank = 0; rank < num_ranks_; rank++) {
+        RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+        auto& ann_if = ann_interfaces_.emplace_back();
+        ann_if.build(dev_resources_[rank], index_params, index_dataset);
+      }
+    } else if (mode_ == SHARDING) {
+      IdxT n_rows           = index_dataset.extent(0);
+      IdxT n_cols           = index_dataset.extent(1);
+      IdxT n_rows_per_shard = (n_rows + num_ranks_ - 1) / num_ranks_;
+      IdxT offset           = 0;
+      for (int rank = 0; rank < num_ranks_; rank++) {
+        RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+        n_rows_per_shard       = std::min(n_rows_per_shard, n_rows - offset);
+        const T* partition_ptr = index_dataset.data_handle() + (offset * n_cols);
+        auto partition         = raft::make_host_matrix_view<const T, IdxT, row_major>(
+          partition_ptr, n_rows_per_shard, n_cols);
```

Comment:
Is it really the job of `ann_mg_index` to shard the data? An alternative is to provide this sharding code as a helper function, that the user has to explicitly call to provide already sharded input to `ann_mg_index::build`

On single node, it make sense to add sharding inside `build`. The hole dataset can be loaded (`mmap`-ed) by the host, and we can shard the data here.

On multiple node the data has to be loaded by each node. Due to its large size, each node would load only their portion of the data. If we allow `ann_mp_index::build` to take that sharded data, then we can already support multi node.

(Yes, I know that multi note is out of scope of the current PR, but if we plan to add that in the future, then we should probably agree already now on a build API that would easily enables that).